### DISCUSSION TITLE
How to quit py5 (with IDLE 3.13)

### DISCUSSION THREAD

**Discussion by wissme** [py5coding/py5generator#562] [Q&A]
Hi ! Just beginning to use py5, very nice system. Two questions :
a) The graphics window pops behind the shell window ??? Obliged to move the shell window :-(
b) Why this program does not quit, I must "restart shell" manually to get the toplevel prompt (IDLE python 3.13).
```
import py5

def setup():
    py5.size(100,30)
    global x,y
    (x,y) = (0,15)
    
def draw():
    global x
    x = x + 1
    py5.ellipse(x,y,10,10)
    if x > 100 : py5.no_loop()

py5.run_sketch()

```

**Comment by hx2A** [comment]
Hi @wissme ! Let me see what I can do to help you.

I was able to reproduce problem a). A simple workaround is to call `py5.get_surface().set_always_on_top(True)` in `setup()`. If you don't actually want it to be always on top, you can later call `py5.get_surface().set_always_on_top(False)`, but not in `setup()`. Alternatively, you can use the `P2D` renderer.

I looked at the code and I know this is something I addressed before but for some reason this problem has returned. Fixing it elegantly will be a bit of a pain but I'll figure it out. Processing in general has problems with MacOS computers.

I cannot reproduce b), but it isn't totally clear to me what you are doing. Can you elaborate? How are you running the program? When you say you cannot exit, do you mean you cannot close the Sketch window or you cannot exit Python?

**Reply by wissme** [reply]
I cannot go back to the toplevel, I must "restart shell" to get the Python prompt >>>

**Reply by hx2A** [reply]
Can you elaborate here? I don't follow what you mean by "toplevel".

**Reply by wissme** [reply]
toplevel == python shell (REPL)

**Reply by wissme** [reply]
With P2D renderer I get a fatal crash of Python.

**Reply by hx2A** [reply]
> toplevel == python shell (REPL)

When you use the generic python interpreter, py5 will by default "block" on the `py5.run_sketch()` call. You won't be able to use the interpreter until after you exit the Sketch. I think on MacOS `py5.run_sketch(block=True)` is required? I forget at the moment. You can try setting it to `False` and see if it lets you.

In any case, can you try using the IPython interpreter instead? That's what I normally use, it's a lot nicer than the generic python interpreter.

**Reply by hx2A** [reply]
> With P2D renderer I get a fatal crash of Python.

Wait, what? That shouldn't happen. Can you elaborate? I'd like to explore this further.

**Comment by hx2A** [comment]
Oh, for `size()` params the minimum size is 100,100. I don't think it will let you create a window smaller than that.

**Reply by wissme** [reply]
yes it creates the window, it's the same with a big window.

**Comment by hx2A** [comment]
Just read your comment in the other discussion thread about problem a) and saw that I wrote about this in August. I said I would open an issue for this but didn't do it. Let me do that now.

**Reply by wissme** [reply]
Ok. Thanks.
